task_3/counter.cpp

```diff
@@ -0,0 +1,324 @@
+#include "counter.h"
+
+#include <iostream>
+#include <boost/bind.hpp>
+#include <boost/asio.hpp>
+#include <boost/thread/thread.hpp>
+#include <experimental/filesystem>
+#include <chrono>
+#include <queue>
+#include <fstream>
+
+using std::string;
+using std::cout;
+using std::endl;
+using namespace std::experimental::filesystem;
+
+boost::mutex global_stream_lock;
+
+	Counter::Counter(path dir, string outFile) : _dir {dir}, _outFile {outFile}, finishedSearching {false}, bunchFilesToPush {2}
+		, bunchFilesToProcess {2} {	}
+
+	void Counter::searchFiles() {
+		global_stream_lock.lock();
+		std::cout << "[" << boost::this_thread::get_id() << "] searchFiles thread" << std::endl;
+		global_stream_lock.unlock();
+
+		std::queue<path> _files;
+		int n = 0;
+		// cycle analizing input directory and pushing decent files to queue
+		for(auto& p: std::experimental::filesystem::recursive_directory_iterator(_dir)) {
+			if (std::experimental::filesystem::is_regular_file(p) &&
+				(path(p).extension() == ".c" || path(p).extension() == ".cpp" ||
+				 path(p).extension() == ".h" || path(p).extension() == ".hpp") )
+        		{
+				_files.push(p);
+				++n;
+				if (n == bunchFilesToPush) {
+					global_stream_lock.lock();	// relocate files from [_files] to [files]
+					for (size_t i = 0; i < bunchFilesToPush; ++i) {
+						files.push(_files.front());
+						cout << "[" << boost::this_thread::get_id() << "] ";
+						cout << "pushed file " << _files.front() << endl;
+						_files.pop();
+					}
+					n = 0;
+					global_stream_lock.unlock();
+				}
+			}
+		}
+		if (_files.size() != 0) {				// relocate rest files from [_files] to [files]
+			global_stream_lock.lock();	
+			while (_files.size() != 0) {
+				files.push(_files.front());
+				_files.pop();
+			}
+			global_stream_lock.unlock();
+		}
+		global_stream_lock.lock();
+		finishedSearching = true;
+		/*cout << "[" << boost::this_thread::get_id() << "] searchFiles thread";
+		cout << "files.size() " << files.size() << ", finishedSearching:" << finishedSearching << endl;*/
+		global_stream_lock.unlock();
+	}
+	
+	void Counter::countStatistic() {
+		global_stream_lock.lock();
+		cout << "[" << boost::this_thread::get_id() << "] countStatistic thread" << endl;
+		global_stream_lock.unlock();
+		
+		std::queue<path> _files;	
+		Statistics _statistic;
+
+		while (!finishedSearching /*&& files.size()*/) {
```

Comment:
Comment this out. From code it is unclear what exactly **finishedSearching** variable does. Try to avoid this kind of hanging relations further.

Reply:
done